pkg/controller/test-images-distributor/test_images_distributor.go

```diff
@@ -0,0 +1,364 @@
+		registryClient:           imagestreamtagwrapper.MustNew(registryManager.GetClient(), registryManager.GetCache()),
+		buildClusterClients:      map[string]ctrlruntimeclient.Client{},
+		pullSecretGetter:         pullSecretGetter,
+		successfulImportsCounter: successfulImportsCounter,
+		failedImportsCounter:     failedImportsCounter,
+		dryRun:                   dryRun,
+	}
+	c, err := controller.New(controllerName, mgr, controller.Options{
+		Reconciler:              r,
+		MaxConcurrentReconciles: 1,
+	})
+	if err != nil {
+		return fmt.Errorf("failed to construct controller: %w", err)
+	}
+
+	buildClusters := sets.String{}
+	for buildClusterName, buildClusterManager := range buildClusterManagers {
+		buildClusters.Insert(buildClusterName)
+		r.buildClusterClients[buildClusterName] = imagestreamtagwrapper.MustNew(buildClusterManager.GetClient(), buildClusterManager.GetCache())
+		// TODO: Watch buildCluster ImageStreams as well. For now we assume no one will tamper with them.
+	}
+
+	objectFilter, err := testInputImageStreamTagFilterFactory(log, configAgent)
+	if err != nil {
+		return fmt.Errorf("failed to get filter for ImageStreamTags: %w", err)
+	}
+	if err := c.Watch(
+		source.NewKindWithCache(&imagev1.ImageStream{}, registryManager.GetCache()),
+		registryClusterHandlerFactory(buildClusters, objectFilter),
+	); err != nil {
+		return fmt.Errorf("failed to create watch for ImageStreams: %w", err)
+	}
+
+	r.log.Info("Successfully added reconciler to manager")
+	return nil
+}
+
+type objectFilter func(types.NamespacedName) bool
+
+// registryClusterHandlerFactory produces a handler that:
+// * Watches ImageStreams because ImageStreamTags do not support the watch verb
+// * Extracts all ImageStramTags out of the Image
+// * Filters out the ones that are not in use
+// Note: We can not use a predicate because that is directly applied on the source and the source yields ImageStreams, not ImageStreamTags
+// * Creates a reconcile.Request per cluster and ImageStreamTag
+func registryClusterHandlerFactory(buildClusters sets.String, filter objectFilter) handler.EventHandler {
+	return &handler.EnqueueRequestsFromMapFunc{ToRequests: handler.ToRequestsFunc(
+		func(mo handler.MapObject) []reconcile.Request {
+			imageStream, ok := mo.Object.(*imagev1.ImageStream)
+			if !ok {
+				logrus.WithField("type", fmt.Sprintf("%T", mo.Object)).Error("Got object that was not an ImageStram")
+				return nil
+			}
+			var requests []reconcile.Request
+			for _, imageStreamTag := range imageStream.Spec.Tags {
+				// Not sure why this happens but seems to be a thing
+				if imageStreamTag.Name == "" {
+					continue
+				}
+				name := types.NamespacedName{
+					Namespace: mo.Meta.GetNamespace(),
+					Name:      fmt.Sprintf("%s:%s", mo.Meta.GetName(), imageStreamTag.Name),
+				}
+				if !filter(name) {
+					continue
+				}
+				// We have to squeeze both the target cluster name and the imageStreamTag name into a reconcile.Request
```

Comment:
Instead of a workqueue per build cluster?

Reply:
yeah, that would be the alternate. I guess my reason for doing it like this is that I think its good if we can limit the total number of parallel imports by setting the worker count. In terms of complexity I would assume doing this vs creating multiple controllers at runtime is probably about the same.